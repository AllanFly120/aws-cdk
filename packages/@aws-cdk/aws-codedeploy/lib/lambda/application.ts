import cdk = require("@aws-cdk/cdk");
import { CfnApplication } from "../codedeploy.generated";
import { ComputePlatform } from "../config";
import { applicationNameToArn } from "../utils";

/**
 * Represents a reference to a CodeDeploy Application deploying to AWS Lambda.
 *
 * If you're managing the Application alongside the rest of your CDK resources,
 * use the {@link LambdaApplication} class.
 *
 * If you want to reference an already existing Application,
 * or one defined in a different CDK Stack,
 * use the {@link #import} method.
 */
export interface ILambdaApplication {
  readonly applicationArn: string;
  readonly applicationName: string;
  readonly computePlatform: ComputePlatform.Lambda;

  export(): LambdaApplicationImportProps;
}

/**
 * Construction properties for {@link LambdaApplication}.
 */
export interface LambdaApplicationProps {
  /**
   * The physical, human-readable name of the CodeDeploy Application.
   *
   * @default an auto-generated name will be used
   */
  applicationName?: string;
}

/**
 * A CodeDeploy Application that deploys to an AWS Lambda function.
 */
export class LambdaApplication extends cdk.Construct implements ILambdaApplication {
  /**
   * Import an Application defined either outside the CDK,
   * or in a different CDK Stack and exported using the {@link #export} method.
   *
   * @param parent the parent Construct for this new Construct
   * @param id the logical ID of this new Construct
   * @param props the properties of the referenced Application
   * @returns a Construct representing a reference to an existing Application
   */
  public static import(scope: cdk.Construct, id: string, props: LambdaApplicationImportProps): ILambdaApplication {
    return new ImportedLambdaApplication(scope, id, props);
  }

  public readonly applicationArn: string;
  public readonly applicationName: string;

  public readonly computePlatform: ComputePlatform.Lambda;

  constructor(scope: cdk.Construct, id: string, props: LambdaApplicationProps = {}) {
    super(scope, id);
    const resource = new CfnApplication(this, 'Resource', {
      applicationName: props.applicationName,
      computePlatform: ComputePlatform.Lambda,
    });

    this.computePlatform = ComputePlatform.Lambda;
    this.applicationName = resource.ref;
    this.applicationArn = applicationNameToArn(this.applicationName, this);
  }

  public export(): LambdaApplicationImportProps {
    return {
      applicationName: new cdk.Output(this, 'ApplicationName', { value: this.applicationName }).makeImportValue().toString(),
      computePlatform: this.computePlatform
    };
  }
}

/**
 * Properties of a reference to a CodeDeploy Application.
 *
 * @see LambdaApplication#import
 * @see ILambdaApplication#export
 */
export interface LambdaApplicationImportProps {
  /**
   * The physical, human-readable name of the Lambda Application we're referencing.
   * The Application must be in the same account and region as the root Stack.
   */
  applicationName: string;

  /**
   * The compute platform of the application, is always Lambda.
   */
  computePlatform?: ComputePlatform.Lambda;
}

class ImportedLambdaApplication extends cdk.Construct implements ILambdaApplication {
  public readonly applicationArn: string;
  public readonly applicationName: string;
  public readonly computePlatform: any;
  constructor(scope: cdk.Construct, id: string, private readonly props: LambdaApplicationImportProps) {
    super(scope, id);
    this.applicationName = props.applicationName;
    this.applicationArn = applicationNameToArn(props.applicationName, this);
    this.computePlatform = ComputePlatform.Lambda;
  }
  public export(): LambdaApplicationImportProps {
    return this.props;
  }
}
