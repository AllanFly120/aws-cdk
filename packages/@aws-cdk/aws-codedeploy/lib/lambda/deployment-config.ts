/**
 * Deployment Config for Lambda CodeDeploy Applications.
 */
export enum LambdaDeploymentConfig {
  AllAtOnce = 'AllAtOnce',
  Canary10Percent30Minutes = 'Canary10Percent30Minutes',
  Canary10Percent5Minutes = 'Canary10Percent5Minutes',
  Canary10Percent10Minutes = 'Canary10Percent10Minutes',
  Canary10Percent15Minutes = 'Canary10Percent15Minutes',
  Linear10PercentEvery10Minutes = 'Linear10PercentEvery10Minutes',
  Linear10PercentEvery1Minute = 'Linear10PercentEvery1Minute',
  Linear10PercentEvery2Minutes = 'Linear10PercentEvery2Minutes',
  Linear10PercentEvery3Minutes = 'Linear10PercentEvery3Minutes'
}