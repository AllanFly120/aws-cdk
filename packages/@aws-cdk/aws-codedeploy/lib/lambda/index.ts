export * from './application';
export * from './deployment-group';
