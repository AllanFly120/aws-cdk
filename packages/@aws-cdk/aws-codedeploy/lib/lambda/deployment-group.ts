import cloudwatch = require('@aws-cdk/aws-cloudwatch');
import iam = require('@aws-cdk/aws-iam');
import lambda = require('@aws-cdk/aws-lambda');
import cdk = require('@aws-cdk/cdk');

import { CfnDeploymentGroup } from '../codedeploy.generated';
import { AutoRollbackConfig, DeploymentOption, DeploymentType } from '../config';
import { deploymentGroupNameToArn, renderAlarmConfiguration, renderAutoRollbackConfiguration } from '../utils';
import { ILambdaApplication, LambdaApplication } from './application';
import { LambdaDeploymentConfig } from './deployment-config';

export interface ILambdaDeploymentGroup extends cdk.IConstruct {
  readonly application: ILambdaApplication;
  readonly deploymentGroupName: string;
  readonly deploymentGroupArn: string;

  export(): LambdaDeploymentGroupImportProps;
}

export interface LambdaDeploymentGroupProps {
  application: LambdaApplication;
  deploymentGroupName?: string;
  deploymentConfig: LambdaDeploymentConfig;

  alarms?: cloudwatch.Alarm[];

  serviceRole?: iam.Role;
  alias: lambda.Alias;

  preHook?: lambda.IFunction;
  postHook?: lambda.IFunction;

  /**
   * Whether to continue a deployment even if fetching the alarm status from CloudWatch failed.
   *
   * @default false
   */
  ignorePollAlarmsFailure?: boolean;

  /**
   * The auto-rollback configuration for this Deployment Group.
   */
  autoRollback?: AutoRollbackConfig;
}
export class LambdaDeploymentGroup extends cdk.Construct implements ILambdaDeploymentGroup {
  /**
   * Import an Lambda Deployment Group defined either outside the CDK,
   * or in a different CDK Stack and exported using the {@link #export} method.
   *
   * @param parent the parent Construct for this new Construct
   * @param id the logical ID of this new Construct
   * @param props the properties of the referenced Deployment Group
   * @returns a Construct representing a reference to an existing Deployment Group
   */
  public static import(scope: cdk.Construct, id: string, props: LambdaDeploymentGroupImportProps): ILambdaDeploymentGroup {
    return new ImportedLambdaDeploymentGroup(scope, id, props);
  }

  public readonly application: ILambdaApplication;
  public readonly deploymentGroupName: string;
  public readonly deploymentGroupArn: string;

  constructor(scope: cdk.Construct, id: string, props: LambdaDeploymentGroupProps) {
    super(scope, id);

    let serviceRole: iam.Role | undefined = props.serviceRole;
    if (serviceRole) {
      if (serviceRole.assumeRolePolicy) {
        serviceRole.assumeRolePolicy.addStatement(new iam.PolicyStatement()
          .addAction('sts:AssumeRole')
          .addServicePrincipal('codedeploy.amazonaws.com'));
      }
    } else {
      serviceRole = new iam.Role(this, 'ServiceRole', {
        assumedBy: new iam.ServicePrincipal('codedeploy.amazonaws.com')
      });
    }
    // Narrow re-implementation of arn:aws:iam::aws:policy/service-role/AWSCodeDeployRoleForLambda
    serviceRole.addToPolicy(new iam.PolicyStatement()
      .addResource('arn:aws:s3:::*/CodeDeploy/*')
      .addActions('s3:GetObject', 's3:GetObjectVersion'));
    serviceRole.addToPolicy(new iam.PolicyStatement()
      .addResource('*')
      .addCondition('StringEquals', {
        's3:ExistingObjectTag/UseWithCodeDeploy': 'true'
      })
      .addActions('s3:GetObject', 's3:GetObjectVersion'));
    serviceRole.addToPolicy(new iam.PolicyStatement()
      .addResource(props.alias.functionArn)
      .addActions('lambda:UpdateAlias', 'lambda:GetAlias'));
    if (props.alarms) {
      serviceRole.addToPolicy(new iam.PolicyStatement()
        .addResources(...props.alarms.map(alarm => alarm.alarmArn))
        .addAction('cloudwatch:DescribeAlarms'));
    }

    const alarms = props.alarms || [];

    const resource = new CfnDeploymentGroup(this, 'Resource', {
      applicationName: props.application.applicationName,
      serviceRoleArn: serviceRole.roleArn,
      deploymentGroupName: props.deploymentGroupName,
      deploymentConfigName: `CodeDeployDefault.Lambda${props.deploymentConfig}`,
      deploymentStyle: {
        deploymentType: DeploymentType.BlueGreen,
        deploymentOption: DeploymentOption.WithTrafficControl
      },

      alarmConfiguration: new cdk.Token(() => renderAlarmConfiguration(alarms, props.ignorePollAlarmsFailure)),
      autoRollbackConfiguration: new cdk.Token(() => renderAutoRollbackConfiguration(alarms, props.autoRollback)),
    });

    this.application = props.application;
    this.deploymentGroupName = resource.deploymentGroupName;
    this.deploymentGroupArn = deploymentGroupNameToArn(this.application.applicationName, this.deploymentGroupName, this);

    if (props.preHook) {
      this.grantPutLifecycleEventHookExecutionStatus(props.preHook.role);
      props.preHook.grantInvoke(serviceRole);
    }
    if (props.postHook) {
      this.grantPutLifecycleEventHookExecutionStatus(props.postHook.role);
      props.postHook.grantInvoke(serviceRole);
    }

    (props.alias.node.findChild('Resource') as lambda.CfnAlias).options.updatePolicy = {
      codeDeployLambdaAliasUpdate: {
        applicationName: props.application.applicationName,
        deploymentGroupName: resource.deploymentGroupName,
        beforeAllowTrafficHook: props.preHook === undefined ? undefined : props.preHook.functionName,
        afterAllowTrafficHook: props.postHook === undefined ? undefined : props.postHook.functionName
      }
    };
  }

  /**
   * Grant a principal permission to codedeploy:PutLifecycleEventHookExecutionStatus
   * on this deployment group resource.
   * @param principal to grant permission to
   */
  public grantPutLifecycleEventHookExecutionStatus(principal?: iam.IPrincipal) {
    if (principal) {
      principal.addToPolicy(new iam.PolicyStatement()
        .addResource(this.deploymentGroupArn)
        .addAction('codedeploy:PutLifecycleEventHookExecutionStatus'));
    }
  }

  public export(): LambdaDeploymentGroupImportProps {
    return {
      application: this.application,
      deploymentGroupName: new cdk.Output(this, 'DeploymentGroupName', {
        value: this.deploymentGroupName
      }).makeImportValue().toString()
    };
  }
}

/**
 * Properties of a reference to a CodeDeploy EC2/on-premise Deployment Group.
 *
 * @see ServerDeploymentGroup#import
 * @see IServerDeploymentGroup#export
 */
export interface LambdaDeploymentGroupImportProps {
  /**
   * The reference to the CodeDeploy EC2/on-premise Application
   * that this Deployment Group belongs to.
   */
  application: ILambdaApplication;

  /**
   * The physical, human-readable name of the CodeDeploy EC2/on-premise Deployment Group
   * that we are referencing.
   */
  deploymentGroupName: string;
}

class ImportedLambdaDeploymentGroup extends cdk.Construct implements ILambdaDeploymentGroup {
  public readonly application: ILambdaApplication;
  public readonly deploymentGroupName: string;
  public readonly deploymentGroupArn: string;

  constructor(scope: cdk.Construct, id: string, private readonly props: LambdaDeploymentGroupImportProps) {
    super(scope, id);
    this.application = props.application;
    this.deploymentGroupName = props.deploymentGroupName;
    this.deploymentGroupArn = deploymentGroupNameToArn(props.application.applicationName,
      props.deploymentGroupName, this);
  }

  public export() {
    return this.props;
  }
}
