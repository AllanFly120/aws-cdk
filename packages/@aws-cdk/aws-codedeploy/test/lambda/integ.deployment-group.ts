import cloudwatch = require('@aws-cdk/aws-cloudwatch');
import lambda = require('@aws-cdk/aws-lambda');
import cdk = require('@aws-cdk/cdk');
import codedeploy = require('../../lib');

const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-codedeploy-server-dg');

function makeAlias(id: string) {
  const func = new lambda.Function(stack, `Func${id}`, {
    code: lambda.Code.inline('exports.handler = function(){ return "test"; };'),
    handler: 'index.handler',
    runtime: lambda.Runtime.NodeJS810
  });
  return new lambda.Alias(stack, `Alias${id}`, {
    aliasName: `alias-${id}`,
    version: new lambda.Version(stack, `Version${id}`, {
      lambda: func
    })
  });
}

const preHook = makeAlias('pre-hook');
const postHook = makeAlias('post-hook');

const application = new codedeploy.LambdaApplication(stack, 'App');
const inPlaceAlias = makeAlias('in-place');
new codedeploy.LambdaDeploymentGroup(stack, 'InPlaceDeployment', {
  alias: inPlaceAlias,
  application,
  deploymentConfig: codedeploy.LambdaDeploymentConfig.AllAtOnce,
  alarms: [
    new cloudwatch.Alarm(stack, 'InPlaceErrors', {
      comparisonOperator: cloudwatch.ComparisonOperator.GreaterThanThreshold,
      threshold: 1,
      evaluationPeriods: 1,
      metric: inPlaceAlias.metricErrors()
    })
  ],
  preHook,
  postHook
});

const blueGreenAlias = makeAlias('blue-green');
new codedeploy.LambdaDeploymentGroup(stack, 'BlueGreenDeployment', {
  alias: blueGreenAlias,
  application,
  deploymentConfig: codedeploy.LambdaDeploymentConfig.Canary10Percent10Minutes,
  alarms: [
    new cloudwatch.Alarm(stack, 'BlueGreenErrors', {
      comparisonOperator: cloudwatch.ComparisonOperator.GreaterThanThreshold,
      threshold: 1,
      evaluationPeriods: 1,
      metric: blueGreenAlias.metricErrors()
    })
  ],
  preHook,
  postHook
});

app.run();
