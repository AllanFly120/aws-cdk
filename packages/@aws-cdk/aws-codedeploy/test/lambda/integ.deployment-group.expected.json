{
  "Resources": {
    "FuncprehookServiceRole6746EC65": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "Funcprehook26EEF433": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = function(){ return \"test\"; };"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "FuncprehookServiceRole6746EC65",
            "Arn"
          ]
        },
        "Runtime": "nodejs8.10"
      },
      "DependsOn": [
        "FuncprehookServiceRole6746EC65"
      ]
    },
    "VersionprehookD4DFA5F7": {
      "Type": "AWS::Lambda::Version",
      "Properties": {
        "FunctionName": {
          "Ref": "Funcprehook26EEF433"
        }
      }
    },
    "Aliasprehook22D81695": {
      "Type": "AWS::Lambda::Alias",
      "Properties": {
        "FunctionName": {
          "Ref": "Funcprehook26EEF433"
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "VersionprehookD4DFA5F7",
            "Version"
          ]
        },
        "Name": "alias-pre-hook"
      }
    },
    "FuncposthookServiceRole0DE6D1C9": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "Funcposthook745F94DB": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = function(){ return \"test\"; };"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "FuncposthookServiceRole0DE6D1C9",
            "Arn"
          ]
        },
        "Runtime": "nodejs8.10"
      },
      "DependsOn": [
        "FuncposthookServiceRole0DE6D1C9"
      ]
    },
    "Versionposthook4F4F6316": {
      "Type": "AWS::Lambda::Version",
      "Properties": {
        "FunctionName": {
          "Ref": "Funcposthook745F94DB"
        }
      }
    },
    "Aliasposthook0ACC00D6": {
      "Type": "AWS::Lambda::Alias",
      "Properties": {
        "FunctionName": {
          "Ref": "Funcposthook745F94DB"
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "Versionposthook4F4F6316",
            "Version"
          ]
        },
        "Name": "alias-post-hook"
      }
    },
    "AppF1B96344": {
      "Type": "AWS::CodeDeploy::Application",
      "Properties": {
        "ComputePlatform": "Lambda"
      }
    },
    "FuncinplaceServiceRoleB3927584": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "FuncinplaceA6026D77": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = function(){ return \"test\"; };"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "FuncinplaceServiceRoleB3927584",
            "Arn"
          ]
        },
        "Runtime": "nodejs8.10"
      },
      "DependsOn": [
        "FuncinplaceServiceRoleB3927584"
      ]
    },
    "Versioninplace239395AD": {
      "Type": "AWS::Lambda::Version",
      "Properties": {
        "FunctionName": {
          "Ref": "FuncinplaceA6026D77"
        }
      }
    },
    "Aliasinplace522E83EC": {
      "Type": "AWS::Lambda::Alias",
      "Properties": {
        "FunctionName": {
          "Ref": "FuncinplaceA6026D77"
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "Versioninplace239395AD",
            "Version"
          ]
        },
        "Name": "alias-in-place"
      },
      "UpdatePolicy": {
        "CodeDeployLambdaAliasUpdate": {
          "ApplicationName": {
            "Ref": "AppF1B96344"
          },
          "DeploymentGroupName": {
            "Ref": "InPlaceDeployment74954D45"
          },
          "BeforeAllowTrafficHook": {
            "Ref": "Aliasprehook22D81695"
          },
          "AfterAllowTrafficHook": {
            "Ref": "Aliasposthook0ACC00D6"
          }
        }
      }
    },
    "InPlaceErrors807304D2": {
      "Type": "AWS::CloudWatch::Alarm",
      "Properties": {
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Threshold": 1,
        "Dimensions": [
          {
            "Name": "FunctionName",
            "Value": {
              "Ref": "Aliasinplace522E83EC"
            }
          }
        ],
        "MetricName": "Errors",
        "Namespace": "AWS/Lambda",
        "Period": 300,
        "Statistic": "Sum"
      }
    },
    "InPlaceDeploymentServiceRole896D55D3": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codedeploy.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "InPlaceDeploymentServiceRoleDefaultPolicy1C2862F9": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject",
                "s3:GetObjectVersion"
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::*/CodeDeploy/*"
            },
            {
              "Action": [
                "s3:GetObject",
                "s3:GetObjectVersion"
              ],
              "Condition": {
                "StringEquals": {
                  "s3:ExistingObjectTag/UseWithCodeDeploy": "true"
                }
              },
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": [
                "lambda:UpdateAlias",
                "lambda:GetAlias"
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "Aliasinplace522E83EC"
              }
            },
            {
              "Action": "cloudwatch:DescribeAlarms",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "InPlaceErrors807304D2",
                  "Arn"
                ]
              }
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Ref": "Aliasprehook22D81695"
              }
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Ref": "Aliasposthook0ACC00D6"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "InPlaceDeploymentServiceRoleDefaultPolicy1C2862F9",
        "Roles": [
          {
            "Ref": "InPlaceDeploymentServiceRole896D55D3"
          }
        ]
      }
    },
    "InPlaceDeployment74954D45": {
      "Type": "AWS::CodeDeploy::DeploymentGroup",
      "Properties": {
        "ApplicationName": {
          "Ref": "AppF1B96344"
        },
        "ServiceRoleArn": {
          "Fn::GetAtt": [
            "InPlaceDeploymentServiceRole896D55D3",
            "Arn"
          ]
        },
        "AlarmConfiguration": {
          "Alarms": [
            {
              "Name": {
                "Ref": "InPlaceErrors807304D2"
              }
            }
          ],
          "Enabled": true
        },
        "AutoRollbackConfiguration": {
          "Enabled": true,
          "Events": [
            "DEPLOYMENT_FAILURE",
            "DEPLOYMENT_STOP_ON_ALARM"
          ]
        },
        "DeploymentConfigName": "CodeDeployDefault.LambdaAllAtOnce",
        "DeploymentStyle": {
          "DeploymentOption": "WITH_TRAFFIC_CONTROL",
          "DeploymentType": "BLUE_GREEN"
        }
      }
    },
    "FuncbluegreenServiceRole6FF59E92": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "Funcbluegreen478A48C3": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "exports.handler = function(){ return \"test\"; };"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "FuncbluegreenServiceRole6FF59E92",
            "Arn"
          ]
        },
        "Runtime": "nodejs8.10"
      },
      "DependsOn": [
        "FuncbluegreenServiceRole6FF59E92"
      ]
    },
    "Versionbluegreen3727C1D4": {
      "Type": "AWS::Lambda::Version",
      "Properties": {
        "FunctionName": {
          "Ref": "Funcbluegreen478A48C3"
        }
      }
    },
    "Aliasbluegreen351D4B71": {
      "Type": "AWS::Lambda::Alias",
      "Properties": {
        "FunctionName": {
          "Ref": "Funcbluegreen478A48C3"
        },
        "FunctionVersion": {
          "Fn::GetAtt": [
            "Versionbluegreen3727C1D4",
            "Version"
          ]
        },
        "Name": "alias-blue-green"
      },
      "UpdatePolicy": {
        "CodeDeployLambdaAliasUpdate": {
          "ApplicationName": {
            "Ref": "AppF1B96344"
          },
          "DeploymentGroupName": {
            "Ref": "BlueGreenDeployment5C188134"
          },
          "BeforeAllowTrafficHook": {
            "Ref": "Aliasprehook22D81695"
          },
          "AfterAllowTrafficHook": {
            "Ref": "Aliasposthook0ACC00D6"
          }
        }
      }
    },
    "BlueGreenErrors60C27452": {
      "Type": "AWS::CloudWatch::Alarm",
      "Properties": {
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Threshold": 1,
        "Dimensions": [
          {
            "Name": "FunctionName",
            "Value": {
              "Ref": "Aliasbluegreen351D4B71"
            }
          }
        ],
        "MetricName": "Errors",
        "Namespace": "AWS/Lambda",
        "Period": 300,
        "Statistic": "Sum"
      }
    },
    "BlueGreenDeploymentServiceRole225851FB": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "codedeploy.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        }
      }
    },
    "BlueGreenDeploymentServiceRoleDefaultPolicy7008FB0A": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject",
                "s3:GetObjectVersion"
              ],
              "Effect": "Allow",
              "Resource": "arn:aws:s3:::*/CodeDeploy/*"
            },
            {
              "Action": [
                "s3:GetObject",
                "s3:GetObjectVersion"
              ],
              "Condition": {
                "StringEquals": {
                  "s3:ExistingObjectTag/UseWithCodeDeploy": "true"
                }
              },
              "Effect": "Allow",
              "Resource": "*"
            },
            {
              "Action": [
                "lambda:UpdateAlias",
                "lambda:GetAlias"
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "Aliasbluegreen351D4B71"
              }
            },
            {
              "Action": "cloudwatch:DescribeAlarms",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "BlueGreenErrors60C27452",
                  "Arn"
                ]
              }
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Ref": "Aliasprehook22D81695"
              }
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Ref": "Aliasposthook0ACC00D6"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "BlueGreenDeploymentServiceRoleDefaultPolicy7008FB0A",
        "Roles": [
          {
            "Ref": "BlueGreenDeploymentServiceRole225851FB"
          }
        ]
      }
    },
    "BlueGreenDeployment5C188134": {
      "Type": "AWS::CodeDeploy::DeploymentGroup",
      "Properties": {
        "ApplicationName": {
          "Ref": "AppF1B96344"
        },
        "ServiceRoleArn": {
          "Fn::GetAtt": [
            "BlueGreenDeploymentServiceRole225851FB",
            "Arn"
          ]
        },
        "AlarmConfiguration": {
          "Alarms": [
            {
              "Name": {
                "Ref": "BlueGreenErrors60C27452"
              }
            }
          ],
          "Enabled": true
        },
        "AutoRollbackConfiguration": {
          "Enabled": true,
          "Events": [
            "DEPLOYMENT_FAILURE",
            "DEPLOYMENT_STOP_ON_ALARM"
          ]
        },
        "DeploymentConfigName": "CodeDeployDefault.LambdaCanary10Percent10Minutes",
        "DeploymentStyle": {
          "DeploymentOption": "WITH_TRAFFIC_CONTROL",
          "DeploymentType": "BLUE_GREEN"
        }
      }
    }
  }
}
